garak/data/__init__.py

```diff
@@ -0,0 +1,60 @@
+# SPDX-FileCopyrightText: Portions Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+# SPDX-License-Identifier: Apache-2.0
+
+"""Local read only resources found by precedence matching supported paths
+
+Ideal usage:
+
+```
+file_path = resources / "filename"
+with open(file_path) as f:
+    f.read()
+```
+
+Resources that do not have a `shipped` version should wrap path access in a try block:
+```
+try:
+    file_path = resources / "filename"
+except GarakException as e:
+    logging.warn("No resource file found.", exc_info=e)
+```
+"""
+
+import pathlib
+
+from garak import _config
+from garak.exception import GarakException
+
+
+class LocalDataPath(pathlib.Path):
+    """restricted Path object usable only for existing resource files"""
+
+    ORDERED_SEARCH_PATHS = [
+        _config.transient.data_dir / "data",
+        _config.transient.package_dir / "data",
+    ]
+
+    def joinpath(self, *pathsegments):
+
+        for segment in pathsegments:
+            prefix_removed = None
+            for path in self.ORDERED_SEARCH_PATHS:
+                if (path == self and segment != "..") or path in self.parents:
+                    prefix_removed = self.relative_to(path)
+                    break
+            if prefix_removed is None:
+                raise GarakException(
```

Comment:
Why not `FileNotFoundError`?

Reply:
Maybe it makes sense to add a garak-specific FNFE to `garak.exceptions`? `GarakPathNotFound` or something?

Reply:
For this error, the actual cause here is more akin to a `ValueError` as the raise is due to a relative path traversal  that is not allowed even if it does exists.

I used a `garak` specific exceptions as consumers of these objects `know` their file exists. To me this means an exception from this method indicates something else `caused` the file to be `removed`, which is different than `not found` IMO.